soroban-sdk/src/account.rs

```diff
@@ -0,0 +1,176 @@
+    }
+}
+
+impl TryFromVal<Env, Object> for Account {
+    type Error = ConversionError;
+
+    fn try_from_val(env: &Env, val: Object) -> Result<Self, Self::Error> {
+        if val.is_obj_type(ScObjectType::Account) {
+            Ok(Account(val.in_env(env)))
+        } else {
+            Err(ConversionError {})
+        }
+    }
+}
+
+impl TryIntoVal<Env, Account> for Object {
+    type Error = <Account as TryFromVal<Env, Object>>::Error;
+
+    fn try_into_val(self, env: &Env) -> Result<Account, Self::Error> {
+        <_ as TryFromVal<_, Object>>::try_from_val(env, self)
+    }
+}
+
+impl TryFromVal<Env, RawVal> for Account {
+    type Error = <Account as TryFromVal<Env, Object>>::Error;
+
+    fn try_from_val(env: &Env, val: RawVal) -> Result<Self, Self::Error> {
+        <_ as TryFromVal<_, Object>>::try_from_val(env, val.try_into()?)
+    }
+}
+
+impl TryIntoVal<Env, Account> for RawVal {
+    type Error = <Account as TryFromVal<Env, Object>>::Error;
+
+    fn try_into_val(self, env: &Env) -> Result<Account, Self::Error> {
+        <_ as TryFromVal<_, RawVal>>::try_from_val(env, self)
+    }
+}
+
+impl IntoVal<Env, Object> for Account {
+    fn into_val(self, _env: &Env) -> Object {
+        self.to_object()
+    }
+}
+
+impl IntoVal<Env, Object> for &Account {
+    fn into_val(self, _env: &Env) -> Object {
+        self.to_object()
+    }
+}
+
+impl IntoVal<Env, RawVal> for Account {
+    fn into_val(self, _env: &Env) -> RawVal {
+        self.to_raw()
+    }
+}
+
+impl IntoVal<Env, RawVal> for &Account {
+    fn into_val(self, _env: &Env) -> RawVal {
+        self.to_raw()
+    }
+}
+
+#[cfg(not(target_family = "wasm"))]
+impl TryFrom<&Account> for ScVal {
+    type Error = ConversionError;
+    fn try_from(v: &Account) -> Result<Self, Self::Error> {
+        ScVal::try_from_val(&v.0.env, v.0.val.to_raw())
+    }
+}
+
+#[cfg(not(target_family = "wasm"))]
+impl TryFrom<Account> for ScVal {
+    type Error = ConversionError;
+    fn try_from(v: Account) -> Result<Self, Self::Error> {
+        (&v).try_into()
+    }
+}
+
+#[cfg(not(target_family = "wasm"))]
+impl TryFromVal<Env, ScVal> for Account {
+    type Error = ConversionError;
+    fn try_from_val(env: &Env, val: ScVal) -> Result<Self, Self::Error> {
+        <_ as TryFromVal<_, Object>>::try_from_val(
+            env,
+            val.try_into_val(env).map_err(|_| ConversionError)?,
+        )
+    }
+}
+
+#[cfg(not(target_family = "wasm"))]
+impl TryIntoVal<Env, Account> for ScVal {
+    type Error = ConversionError;
+    fn try_into_val(self, env: &Env) -> Result<Account, Self::Error> {
+        Account::try_from_val(env, self)
+    }
+}
+
+impl Account {
+    pub(crate) unsafe fn unchecked_new(obj: EnvObj) -> Self {
+        Self(obj)
+    }
+
+    pub fn env(&self) -> &Env {
+        self.0.env()
+    }
+
+    pub fn address(&self) -> Address {
+        self.0.env().get_account_address(&self)
+    }
+
+    pub fn authorize(&self, args: Vec<RawVal>) {
```

Comment:
What does this function do? It is unclear.

Reply:
Sorry, I spend much time on docs, given that this is a prototype/iteration. That's the main authorization function: it authorizes a call of the current contract function with the given args and with the current authorized call stack (see examples). More strictly, calling `authorize` requires the account to have an invocation in signature payload that consists of the call stack of all the  `(contract_id, fn_name)` pairs that have called `authorize` for this account, and `args` provided by the current contract.
Basically, if you want to make an account to have authorized your contract's invocation, you call this. That's all the contract needs to do for auth. Everything else is on host/account contract.

Reply:
I'm still catching up on how this works so I might be way off, and I suspect I'm misunderstanding how this works, so please bear with me!

It's a little confusing that this function returns nothing. If the function is just loading up the authorizations that were in the invocation, why can't the host do that automatically when it receives the invocation?
Otherwise, what if a contract forgets to call it, or is tricked into not calling it. And if it is always the first thing a contract calls, we could just do it host side before calling any WASM.

If there was a function to call for verifying auth, I expect it would give me what was authenticated, like the identity, or ID, or something.

Reply:
> It's a little confusing that this function returns nothing. 

Currently it would just panic if authorization fails in enforcing mode. It could maybe return `bool/Result`, but I'm not convinced if it's ever a good idea to gracefully handle the authorization failure...

> Otherwise, what if a contract forgets to call it, 

Then no authorization happens. That's up to the contract basically. Nothing prevents the contract from e.g. passing just an `Address` instead of `Account` etc. I'm not sure if it's possible to make this completely fool-proof without seriously compromising flexibility.

> And if it is always the first thing a contract calls, we could just do it host side before calling any WASM

We don't know the args beforehand and we may not even always want to call this and consume nonce. We may add some SDK sugar (maybe with some host backing) for the simple cases where there is just a single account for which we need to authorize the invocation with all the function args. This draft currently only implements the generic case that allows passing the accounts through and passing the custom args to `authorize` (see atomic swap/multiswap examples).

Something like the previous roles concept is not out of the question, but I'm not sure how to make it non-confusing.

Reply:
>  I'm not convinced if it's ever a good idea to gracefully handle the authorization failure.

I've just realized that `try_call` kind of allows that anyway. So maybe we should really make this return result or add `try_authorize`... I'll think more about that.